extension/src/ALObject/ALParser.ts

```diff
@@ -0,0 +1,310 @@
+    }
+    control.startLineIndex = control.endLineIndex = lineIndex;
+    control.alCodeLines = parent.alCodeLines;
+    control.parent = parent;
+    return control;
+
+
+}
+
+function getProperty(parent: ALControl, lineIndex: number, codeLine: ALCodeLine) {
+    let propertyResult = codeLine.code.match(/^\s*(?<name>ObsoleteState|SourceTable|PageType)\s*=\s*(?<value>[a-zA-Z]*);/i);
+    if (propertyResult && propertyResult.groups) {
+        let property = new ALProperty(parent, lineIndex, propertyResult.groups.name, propertyResult.groups.value);
+        return property;
+    }
+    return;
+}
+
+function matchIndentationDecreased(codeLine: ALCodeLine): boolean {
+    const indentationDecrease = /(^\s*}|}\s*\/{2}(.*)$|^\s*\bend\b)/i;
+    let decreaseResult = codeLine.code.match(indentationDecrease);
+    return null !== decreaseResult;
+}
+
+function matchIndentationIncreased(codeLine: ALCodeLine): boolean {
+    const indentationIncrease = /^\s*{|{\s*\/{2}(.*)$|\bbegin\b\s*$|\bbegin\b\s*\/{2}(.*)$|\bcase\b\s.*\s\bof\b/i;
+    let increaseResult = codeLine.code.match(indentationIncrease);
+    return null !== increaseResult;
+}
+
+
+function matchLabel(line: string): RegExpExecArray | null {
+    const labelTokenPattern = /^\s*(?<name>\w*): Label (?<text>('(?<text1>[^']*'{2}[^']*)*')|'(?<text2>[^']*)')(?<maxLength3>,\s?MaxLength\s?=\s?(?<maxLengthValue3>\d*))?(?<locked>,\s?Locked\s?=\s?(?<lockedValue>true|false))?(?<maxLength2>,\s?MaxLength\s?=\s?(?<maxLengthValue2>\d*))?(?<comment>,\s?Comment\s?=\s?(?<commentText>('(?<commentText1>[^']*'{2}[^']*)*')|'(?<commentText2>[^']*)'))?(?<locked2>,\s?Locked\s?=\s?(?<lockedValue2>true|false))?(?<maxLength>,\s?MaxLength\s?=\s?(?<maxLengthValue>\d*))?(?<locked3>,\s?Locked\s?=\s?(?<lockedValue3>true|false))?/i;
+    let labelTokenResult = labelTokenPattern.exec(line);
+    return labelTokenResult;
+}
+export function getLabel(parent: ALControl, lineIndex: number, codeLine: ALCodeLine): MultiLanguageObject | undefined {
+    let matchResult = matchLabel(codeLine.code);
+    let mlObject = getMlObjectFromMatch(parent, lineIndex, MultiLanguageType.Label, matchResult);
+    return mlObject;
+}
+
+
+function matchMlProperty(line: string): RegExpExecArray | null {
+    const mlTokenPattern = /^\s*(?<name>OptionCaption|Caption|ToolTip|InstructionalText|PromotedActionCategories|RequestFilterHeading) = (?<text>('(?<text1>[^']*'{2}[^']*)*')|'(?<text2>[^']*)')(?<maxLength3>,\s?MaxLength\s?=\s?(?<maxLengthValue3>\d*))?(?<locked>,\s?Locked\s?=\s?(?<lockedValue>true|false))?(?<maxLength2>,\s?MaxLength\s?=\s?(?<maxLengthValue2>\d*))?(?<comment>,\s?Comment\s?=\s?(?<commentText>('(?<commentText1>[^']*'{2}[^']*)*')|'(?<commentText2>[^']*)'))?(?<locked2>,\s?Locked\s?=\s?(?<lockedValue2>true|false))?(?<maxLength>,\s?MaxLength\s?=\s?(?<maxLengthValue>\d*))?(?<locked3>,\s?Locked\s?=\s?(?<lockedValue3>true|false))?/i;
```

Comment:
In my opinion RegExp with a high number of possible executions should be precompiled. Altough I do not know when it becomes beneficial so I guess it's a pretty baseless opinion.

Also, I do like the idea of storing the expression in another file or in an object. How do you feel about that?

Reply:
I was thinking about moving the matching (including the regex strings) of each object to the corresponding object.
So function matchALControl would move to ALControl, functions matchMlProperty/getMlProperty  and matchLabel/getLabel to MultiLanguageObject etc. Would be easier to maintain and extemd, I guess.

I just happen to like the /[regex]/[switches] syntax, don't know how bad it is compared to precompiled though in matter of performance...

Reply:
I cannot move the functions from ALParser into the classes, I get into circular references with the imports.. :-/
(see more at https://stackoverflow.com/questions/48603541/typeerror-class-extends-value-undefined-is-not-a-constructor-or-null)
So for now this has to be the structure, can restructure later if we find a good solution...

Reply:
reg. RegExp('REGEX string') vs /REGEX string/ (literal syntax)

I don't like RegExp(string) since it'll need some double escaping making the regex text hard to read. The only advantage is that it accepts variable input, which is used in some places in the code.

I think that literal can be faster in some circumstances, but often is equally fast, if I don't misinterpret this: https://stackoverflow.com/a/32523333

So I think we'll be using both approaches, depending on scenario.